Add lightning-on-taproot review to bitcoinplusplus/layer-2 (#640)
bitcointranscripts/bitcointranscripts: bitcoinplusplus/layer-2/lightning-on-taproot.md

---
title: 'Lightning on Taproot'
transcript_by: 'garryk1612-cyber via review.btctranscripts.com'
media: 'https://www.youtube.com/watch?v=E_z4hjvVzoQ'
date: '2023-07-18'
tags:
  - 'lightning'
  - 'taproot'
speakers:
  - 'Arik Sosman'
categories:
  - 'privacy-enhancements'
---
## Introduction to `Taproot` Channls

> **Arik Sosman:**  00:00:04

Reviewer: ```suggestion
**Arik Sosman:**  00:00:04

```
there is always a line between the speaker and the next  paragraph

---
title: 'Lightning on Taproot'
transcript_by: 'garryk1612-cyber via review.btctranscripts.com'
media: 'https://www.youtube.com/watch?v=E_z4hjvVzoQ'
date: '2023-07-18'
tags:
  - 'lightning'
  - 'taproot'
speakers:
  - 'Arik Sosman'
categories:
  - 'privacy-enhancements'
---
## Introduction to `Taproot` Channls

**Arik Sosman:**  00:00:04

My name is Arik, I work at Spiral and most recently I've been working on adding support for `Taproot` channels.
At least I try to work on that, but I'm getting pulled back into Swift bindings.
And this presentation is supposed to be about why the `Taproot` spec is laid out the way it is.
What are some of the motivations, the constraints, limitations that are driving the design and I wanna really dig deep into the math that some of the vulnerabilities that we're trying to avoid as well as how we're solving those issues.

> So first of all, obviously we know that `Taproot` has not been active for a while, but why are we actually bothering with modifying the way that lightning channels are opened such that we can have lightning channels operate on half-route.

Reviewer 2: Please check `operate on half-route`

Author: ohh I know LN channels operate on Taproot

I guess I should have asked this question before showing this slide, but let me just go back and see if the audience has any suggestions they want to know here.

## Privacy & Security

**Audience** 00:01:18

Is there privacy?

**Arik Sosman:** 00:01:22 

Great.
Yeah, so the biggest reason, of course, is that we have privacy improvements.

> At least, assuming that eventually everybody is going to be using hatred `taproot` addresses.

Reviewer 2: ```suggestion
At least, assuming that eventually everybody is going to be using `Pay-to-Taproot` addresses.
```

Reviewer 2: https://river.com/learn/terms/p/pay-to-taproot-p2tr/
https://bitcointreasuries.net/glossary/pay-to-taproot

> At least, assuming that eventually everybody is going to be using patriot `taproot` addresses.

Reviewer 2: `patriot` ?

As you know, `taproot` is segwit v1.
Segwit was segwit v0, that was our new software mechanism.

> If at some point we decide that we need segwit v2, then all of the old lightening `Taproot` channels that aren't human support yet, that are going to be using SegWit v1, are going to lose their privacy status and they're going to once again start sticking out like a sore thumb.

Reviewer 2: ```suggestion
If at some point we decide that we need segwit v2, then all of the old lightening `Taproot` channels that don't even support yet, that are going to be using SegWit v1, are going to lose their privacy status and they're going to once again start sticking out like a sore thumb.
```
Please check. I'm not sure about this one.

So that is one of the benefits.
We are also able to improve privacy by decorrelating payments.
That is `PTLC`s. and that is a privacy benefit that is actually going to persist even if we have a SegWit V2 or V3, because there the primary issue that we have with `HTLC`s right now is that if we were to have multiple channels that have the same inflight `HTLC` go on-chain, then that hash would be correlatable and we would be able to link the payment chain, link one channel to the other.

> There are hopefully also some security improvements such as directional signatures, but it's still very much a research phase.

Reviewer 2: "directional signatures" ???

So you know, mostly I think the greatest benefit really is privacy.

> There is also something to be said about the fact that with the way that `Taproot` works and the way that Lightning requires a bunch of different spend paths, with `Taproot` Taptrees we are able to make a bunch of those pen paths much cheaper, and we are also able to not reveal the ones that are being unused.

Reviewer 2: "pen paths"  ?

So not great, and really highlights how trivial it is to attack and to extract a private key if you have two signatures that have reused the same nonce.
It's really, like I wish this equation were written out more frequently because I think people understate quite how trivial of an attack this is.
I mean you can write Python code in like two minutes.

> ** Audience:** 00:08:11

Reviewer: ```suggestion
** Audience:**  00:08:11
```

Reviewer 3: Well  why are we adding some extra space here ?

> So Sony got pwned, they were reusing nonces on all of their ECBSA signatures.

Reviewer: ```suggestion
So Sony got pwned, they were reusing nonces on all of their `ECDSA` signatures.
```

So not great, and really highlights how trivial it is to attack and to extract a private key if you have two signatures that have reused the same nonce.
It's really, like I wish this equation were written out more frequently because I think people understate quite how trivial of an attack this is.
I mean you can write Python code in like two minutes.

** Audience:**  00:08:11

So Sony got pwned, they were reusing nonces on all of their `ECDSA` signatures.
It was the same nonce for all the signatures for all the Playstations.

## Issues with MUSIG & Lightning

**Arik Sosman:** 00:08:19

Yeah, that is true.
I believe that was the PlayStation 3, right?
Although for `ECDSA signatures`, so the thing about Schnorr is the math is so simple, it's just additional application.
You don't have any of the `ECDSA` complications, so here you don't even have to think that long about why this attack is trivial.
However, obviously we need to mitigate these attacks, so `MuSIG2` comes to the rescue.
I'm sure all of you have heard about `MuSIG2` many times.

> And essentially, the main thing that `MuSIG2` does, it's like the same approach that eliminates the attack vector for both the naive key aggregation and the non-tree views, is it introduces coefficients.

Reviewer 2: `non-tree views` ???

But if you have some other sorting algorithm, or you just know the order and it's not sorted, but it's just ordered beforehand, that also works.
The important thing is just that you know which public key corresponds to which coefficient and then, say if you're Alice and your pub key also comes first, so your index is 1, because in math, indices don't start at 0.
If Alice's index is 1, then she simply takes her own pubkey, concatenates it with a hash of the sequence of all the pubkeys, which in this case would be first hers, then Bob's.
That thing is hashed, and that hash is concatenated with her own pubkey, then we hash that again, and so, because of elliptic curve photography, we can simply interpret hashes as big integers, we take that hash and simply use that as a big integer that is a coefficient for her public key.
Now, if you have been reading the MuSIG2 protocol, then you will know that this is not actually true, because we have an exception for the second index.
But in the grand scheme of things, it doesn't really matter.
So I think for understanding what problem MuSIG solves and how it does that, this is kind of deep enough.
So I hope I haven't confused anybody yet.

> All right, because next is non-segregation.

Reviewer 2: Is it "nonce aggregation" ?

And with nonces, we really do the same thing.
And one might actually wonder, why, if we're already doing this stuff with keys, Do we even have to bother with non-segregation and adding coefficients to our nonces?
And actually, this is probably a good point where I should poll the audience.
Any guesses as to why we care about nonce coefficients also?
Yeah?

**Audience:** 00:12:47

You could fool people into nonce-reuse, right?

**Arik Sosman:** 00:12:51

You could fool people into nonce, yeah.
That is one thing and you could also construct a nonce adversarily in such a way that you would then be able to unilaterally create a signature with `MuSIG2`.
So it's really pretty much the same thing.
And with `nonce`, the protocol is a little more complicated.
But I really still want to go through it, because one of the innovations with `MuSIG2` is how this protocol can work with just one round trip as opposed to multiple round trips that we had with regular MuSIG.
So the way this works is as follows.

> Each participant generates not one lowercase r that we had here, but they generate two So a music two knots is actually not one knots, but it is two knots.

Reviewer 2: Please check "music two knots is actually not one knots, but it is two knots"

And then we can simplify the equation, and we can extract x.
We can extract the private key.
if I hadn't simplified this, if I had also included the other partial nonce, then we simply would have done this elimination step twice because, we're trying to eliminate multiple variables, but we have sufficient equations to do so.
So why do we care about this?
That is because we don't really trust our counterparty.
Because in a multi-state setup, we don't really worry about the rest of the world attacking us and knowing what our private key is.
We also really want to make sure that each one of the participants in a multi-signature cannot know what the other participant's private key is either.
Because as soon as they do, we have lost the whole benefit of having a multi-sig setup.
So now that we know that these are the steps that we need to do in order to avoid the pitfalls with Schnorr.
We need to use MuSIG.
How then do we transcript to actually creating and signing Lightning commitments?
And before we'll dig into that, we need to first consider what the properties of Lightning commitments are.

> First of all, without SIGHASH, any Prevout, or Covenants, or anything of the sort, we are going to have to rely on asymmetrical commitment transactions, which means that when Alice and Bob open a channel with one another, their commitment transactions are not going to look identical because we have the read-sharmony branch and the to-self delay is going to apply to different outputs depending on whether you're looking at the commitment transaction from Alice's or from Bob's perspective.

Reviewer 2: Please check

```suggestion
First of all, without SIGHASH, SIGHASH_ANYPREVOUT, or Covenants, or anything of the sort, we are going to have to rely on asymmetrical commitment transactions, which means that when Alice and Bob open a channel with one another, their commitment transactions are not going to look identical because we have the read-sharmony branch and the to-self delay is going to apply to different outputs depending on whether you're looking at the commitment transaction from Alice's or from Bob's perspective.
```

Reviewer 2: `read-sharmony branch` ?

So right out of the gate, we already have two different messages that we need to sign.
Second, we have the same aggregate key.
Because our aggregate key is our funding key and it doesn't change across each new commitment, it means that we have to be extremely careful not to reuse nonces.

> And because we have to sign multiple commitments, it really means that we have to have two non-spares as opposed to one non-spare that we send out with each exchange.

Reviewer 2: `two non-spares` ?

So let's look at how it would work in practice with, say, a channel opening.
So what you want to get to is you want Alice to have a signed commitment transaction with Bob's private key, such that she can later append her own partial signature where she is in a need to broadcast unilaterally.
We need the same for Bob, too.
If Alice opens a channel to Bob, then of course what she needs is eventually for Bob to provide her with a partial signature in order for Bob to be able to do that, Alice, right out of the gate, needs to send Bob her local nonce and That local nonce, which is the public nonce there, is the one that Bob is going to need in his partial signature, because the commitment incorporates an aggregation of those nonces.
So Bob, in order to be able to calculate the music two-step, needs to know what Alice's local nonce later down the line is going to be.
So, if we just follow the flow, Alice sends Bob her local nonce, as well as the remote nonce such that Bob can create a partial signature for himself.

> We then have Bob respond with the same step where Bob generates the nonce pairs and sends those in the accept channel message and then in Funding Creative, Alice can now provide a partial signature to Bob because what Bob needs is a partial signature that uses Alice's remote nonce, because the conventional transaction that Bob has, from his perspective it is his local, and from Alice's perspective it's Alice's remote.

Reviewer 2: `Funding Creative` ?

So not great, and really highlights how trivial it is to attack and to extract a private key if you have two signatures that have reused the same nonce.
It's really, like I wish this equation were written out more frequently because I think people understate quite how trivial of an attack this is.
I mean you can write Python code in like two minutes.

** Audience:**  00:08:11

So Sony got pwned, they were reusing nonces on all of their `ECDSA` signatures.
It was the same nonce for all the signatures for all the Playstations.

> ## Issues with MUSIG & Lightning

Reviewer 2: ```suggestion
## Issues with MuSig & Lightning
```

Reviewer 2: https://bitcoinops.org/en/topics/musig/

**Arik Sosman:** 00:08:19

Yeah, that is true.
I believe that was the PlayStation 3, right?
Although for `ECDSA signatures`, so the thing about Schnorr is the math is so simple, it's just additional application.
You don't have any of the `ECDSA` complications, so here you don't even have to think that long about why this attack is trivial.

> However, obviously we need to mitigate these attacks, so `MuSIG2` comes to the rescue.

Reviewer 2: ```suggestion
However, obviously we need to mitigate these attacks, so `MuSig2` comes to the rescue.
```

Reviewer 2: Could you please replace all instances of ~MuSIG2~ with `MuSig2`?

Ref: https://bitcoinops.org/en/topics/musig/

The idea is that we're gonna beat the scripts into a manuscript parsers and generators and then see what the optimal script output is.
But just reading this, people might be confused because they'll say, oh, if check safe passed, but say 2-sulfulate is 0, will it go through?
To solve the layout, this should never be 0.
But just reading this, people might be confused because they'll say, oh, if check safe passed, but say to_self_delay is 0, will it go through?
to_self_delay, this should never be 0.
But there's going to be a follow-up case where there's going to be a bigger question.
So for HDLC offers, and I don't really want to go into accepted HDLCs because they're so similar to offered HDLCs. We still have the same situation that the revocation key is the unencumbered spend path.
And for that reason, it becomes the key spend.
So for `HTLC` offers, and I don't really want to go into accepted `HTLC`s because they're so similar to offered `HTLC`s. We still have the same situation that the revocation key is the unencumbered spend path and for that reason, it becomes the key spend.
For script spend paths, We now have two situations.

> So one of them is if we are able to provide the hash preimage, which is marked in green here, And the script looks a little complicated, but the point is, once OPTCHECKSIG verifies here, then it should be 1, and then we have OPTCHECKSQUENCEVERIFY.

Reviewer 2: Could you double-check the Youtube video around the 25:35 mark? It shows `OP_CHECKSEQUENCEVERIFY`, not OP_CSV

Reviewer 2: <img width="2770" height="1658" alt="Image" src="https://github.com/user-attachments/assets/05acd130-658b-4f40-a994-9f647024a526" />

> So one of them is if we are able to provide the hash preimage, which is marked in green here, And the script looks a little complicated, but the point is, once `OP_CHECKSIG` verifies here, then it should be 1, and then we have `OP_CSV`.

Reviewer 2: ```suggestion
So one of them is if we are able to provide the hash preimage, which is marked in green here, And the script looks a little complicated, but the point is, once `OP_CHECKSIG` verifies here, then it should be 1, and then we have `OP_CHECKSEQUENCEVERIFY`.
```

> But if `OP_CHECKSIG` does not verify, then we have a 0, and then we have 0 `OP_CSV` and The question is, would it then mean that without a valid signature, we would be able to spend it without delay?

Reviewer 2: ```suggestion
But if `OP_CHECKSIG` does not verify, then we have a 0, and then we have 0 `OP_CHECKSEQUENCEVERIFY` and The question is, would it then mean that without a valid signature, we would be able to spend it without delay?
```

> Well, it doesn't, because Ops CSV has some weird consideration where a transaction can never actually be spent with a zero sequence because it's doing greater than and not greater equals.

Reviewer 2: ```suggestion
Well, it doesn't, because OP_CSV has some weird consideration where a transaction can never actually be spent with a zero sequence because it's doing greater than and not greater equals.
```

Reviewer 2: https://www.spark.money/glossary/bitcoin-script#key-opcodes
https://github.com/saving-satoshi/saving-satoshi/issues/1394

He'll send the local nonce for the following iteration of the channel exchange.
But really, there isn't much of a difference between channel opening and channel operation.
Really the most important thing is that you have to have separate nonces for your own local commitment transaction and for the remote commitment transaction and it's something that can be a bit of a pain to keep track of.
One thing I will note is if you think that your node setup is such that you will never ever possibly have to sign a local commitment transaction with your own private key multiple times, then in principle, you don't even need to update your own local nonce, because there would never be a nonce freeze.
However, it is something that is very hard to guarantee, so for the purpose of safety, please ignore I said that.
Well, we have pretty much covered the thing that `Taproot` enables with the nonce communication.
Even though on-chain the footprint is smaller, with the nonces we do have this additional headache that all of you who are implementing Lightning protocols are going to have to be cognizant of.
But that is not the only thing that `Taproot` does.

> `Taproot` also, as I have alluded to earlier, allows us to have Tap trees, where instead of having just one massive, dare I say, inscrutable script, especially with all this up-if, up-else nesting, we can very cleanly divide the spend paths into their separate tab branches.

Reviewer 2: Please check.
```suggestion
`Taproot` also, as I have alluded to earlier, allows us to have Tap trees, where instead of having just one massive, dare I say, inscrutable script, especially with all this OP_IF, OP_ELSE nesting, we can very cleanly divide the spend paths into their separate tab branches.
```

The idea is that we're gonna beat the scripts into a manuscript parsers and generators and then see what the optimal script output is.
But just reading this, people might be confused because they'll say, oh, if check safe passed, but say to circulate is 0, will it go through?
To solve the layout, this should never be 0.
But there's going to be a follow-up case where there's going to be a bigger question.
So for `HTLC` offers, and I don't really want to go into accepted `HTLC`s because they're so similar to offered `HTLC`s. We still have the same situation that the revocation key is the unencumbered spend path and for that reason, it becomes the key spend.
For script spend paths, We now have two situations.
So one of them is if we are able to provide the hash preimage, which is marked in green here, And the script looks a little complicated, but the point is, once `OP_CHECKSIG` verifies here, then it should be 1, and then we have `OP_CSV`.
But if `OP_CHECKSIG` does not verify, then we have a 0, and then we have 0 `OP_CSV` and The question is, would it then mean that without a valid signature, we would be able to spend it without delay?

> Well, it doesn't, because Ops CSV has some weird consideration where a transaction can never actually be spent with a zero sequence because it's doing greater than and not greater equals.

Reviewer 2: OP_CSV  ?

Reviewer 2: Ops CSV => OP_CSV

And guess what Carol needs to know in order to be able to spend Bob's `PTLC`?
She needs to learn what a plus b plus c plus z.
And guess what?
Carol needs to know in order to be able to spend Bob's `PTLC`.
She needs to learn what a plus b plus c plus z is, because then she can subtract her own secret that she received in the onion, and that way she can spend Bob's `PTLC`.
So now the question of course is, I think you guys can figure it out, especially with this slide, why are we bothering with this MuSIG complication here?
Why do we have to have music just to create an adapter signature that is tweaked damaged broken Whatever terminology when I use.

**Audience:** 00:47:08

The same reasons you don't want the nodes private key leads, you don't want the payment point linked as well, if you're not doing the same knots, you know, computations or anything.
Just depending on that payment not to be there in general.

**Arik Sasman:** 00:47:24

what would happen if it were to be ?
What scenario are we trying to avoid?

> **Audience:** 00:47:29

Reviewer: ```suggestion
**Audience:** 00:47:29

```
space between here
